lib/src/types.ts

```diff
@@ -37,7 +38,8 @@ export type Type =
   | ObjectType
   | ArrayType
   | UnionType
-  | ReferenceType;
+  | ReferenceType
+  | IntersectionType;
```

Comment:
`IntersectionType` will also need to be excluded in `ConcreteType` below, which will also require a non-trivial change to [`possibleRootTypes`](https://github.com/airtasker/spot/blob/16f5e964dff888e96dfffafeee75617d28602dc7/lib/src/types.ts#L421) - the intersection type will need to be resolved into single merged `object` type

Reply:
@lfportal would be curious when it would need to resolve into a merged `object`. I might be able to get a better understanding of what would need to be returned

Reply:
Currently the primary use case (though there are others) for `possibleRootTypes` is to infer a discriminator property for union types see [inferDiscriminator](https://github.com/airtasker/spot/blob/16f5e964dff888e96dfffafeee75617d28602dc7/lib/src/types.ts#L453). 

`possibleRootTypes` will unwrap unions and dereference types. Consider:

```ts
interface A { x: "a", a: String; }
interface B { x: "b", b: String; }
interface C { x: "c", c: String; }

type BC = B | C;
type ABC = A | BC;

possibleRootTypes(ABC) === [{ x: "a", a: String; }, { x: "b", b: String; }, { x: "c", c: String; }]
```
A discriminator is then inferred by examining each type and checking for a common property with a unique literal value. Intersection types bring a bit more complexity to this:

```ts
interface A { x: "a", a: String; }
interface B { x: "b", b: String; }
interface C { c: String; }
interface X { x: "c"; }

type ABC = A | B | (C & X)
possibleRootTypes(ABC) === [{ x: "a", a: String; }, { x: "b", b: String; }, (C & X)] // <- current behaviour
```
Even though the type `C & X` does actually have a conforming discriminator `x: "c"` is would currently not be explored by the `inferDiscriminator` function.

After thinking this through some more, I think we have two high level options:
1. Handle intersection types appropriately anywhere `possibleRootTypes` is called
2. Change `possibleRootTypes` to resolve intersection types into an `object` type

Admittedly this issue would have been far simpler to spot if parts of the codebase adopted the exhaustive `switch case` style rather than sometimes checking for specific types and defaulting behaviour otherwise 😢

Reply:
I used the same method used for union types to identify the possibleRoot types. Since `allOf` does not have any discriminators, those were not added to the openAPI specification, but were inferred for when we have unions with intersections.

Given how some of the parsing is set up, the example provided above would not work as is, however it does if you change it to the following when defining the contract. Using the parenthesis throws an "unknown type" error. After the changes the following are valid contracts:
Union with Intersection
```
interface A { x: "a", a: string; }
interface B { x: "b", b: string; }
interface C { c: string; }
interface X { x: "c"; }
type CX = C & X;
type ABC = A | B | CX
```
Union with Union
```
interface A { x: "a", a: string; }
interface B { x: "b", b: string; }
interface C { x: "c"; }
interface X { x: "d"; }
type CX = C | X;
type ABC = A | B | CX
```

Finally, another valid discriminator can be boolean types which are ignored. Before adding those in, I wanted to ask if that is a missing feature and should be added?

Reply:
> Finally, another valid discriminator can be boolean types which are ignored. Before adding those in, I wanted to ask if that is a missing feature and should be added?

Let's keep non-string discriminator types as separate issues if there is demand for them.